rust/relay/src/server.rs

```diff
@@ -1,53 +1,166 @@
     decoder: MessageDecoder<Attribute>,
     encoder: MessageEncoder<Attribute>,
+
+    local_ip4_address: SocketAddrV4,
+    local_ip6_address: SocketAddrV6,
+
+    /// All client allocations, indexed by client's socket address.
+    allocations: HashMap<SocketAddr, Allocation>,
+
+    used_ports: HashSet<u16>,
+    pending_commands: VecDeque<Command>,
+    next_allocation_id: AllocationId,
+
+    rng: R,
+}
+
+/// The commands returned from a [`Server`].
+///
+/// The [`Server`] itself is sans-IO, meaning it is the caller responsibility to cause the side-effects described by these commands.
+#[derive(Debug)]
+pub enum Command {
+    SendMessage {
+        payload: Vec<u8>,
+        recipient: SocketAddr,
+    },
+    /// Listen for traffic on the provided IP addresses.
+    ///
+    /// Any incoming data should be handed to the [`Server`] via [`Server::handle_relay_input`].
+    /// The caller MUST deallocate the port after the given duration unless it is refreshed.
+    AllocateAddresses {
+        id: AllocationId,
+        ip4: SocketAddrV4,
+        ip6: SocketAddrV6,
+        expiry_in: Duration,
+    },
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
+pub struct AllocationId(u64);
+
+impl AllocationId {
+    fn next(&mut self) -> Self {
+        let id = self.0;
+
+        self.0 += 1;
+
+        AllocationId(id)
+    }
 }
 
+/// See <https://www.rfc-editor.org/rfc/rfc8656#name-requested-transport>
+const UDP_TRANSPORT: u8 = 17;
+
+/// The maximum number of ports available for allocation.
+const MAX_AVAILABLE_PORTS: u16 = 65535 - 49152;
+
+/// The maximum lifetime of an allocation.
+const MAX_ALLOCATION_LIFETIME: Duration = Duration::from_secs(3600);
+
+/// The default lifetime of an allocation.
+///
+/// TODO: This has been chosen at random by Thomas. Revisit if it makes sense.
+const DEFAULT_ALLOCATION_LIFETIME: Duration = Duration::from_secs(600);
+
 impl Server {
-    /// Process the bytes received from one node and optionally return bytes to send back to the same or a different node.
-    pub fn handle_received_bytes(
-        &mut self,
-        bytes: &[u8],
-        sender: SocketAddr,
-    ) -> Result<Option<(Vec<u8>, SocketAddr)>> {
+    pub fn new(local_ip4_address: SocketAddrV4, local_ip6_address: SocketAddrV6) -> Self {
+        // TODO: Validate that local IPs aren't multicast / loopback etc.
+
+        Self {
+            decoder: Default::default(),
+            encoder: Default::default(),
+            local_ip4_address,
+            local_ip6_address,
+            allocations: Default::default(),
+            used_ports: Default::default(),
+            pending_commands: Default::default(),
+            next_allocation_id: AllocationId(1),
+            rng: rand::thread_rng(),
+        }
+    }
+}
+
+impl<R> Server<R>
+where
+    R: Rng,
+{
+    /// Process the bytes received from a client.
+    ///
+    /// After calling this method, you should call [`next_event`] until it returns `None`.
```

Comment:
Isn't it `next_command`?

Reply:
Yep! Reminds me that we need a CI check to lint against unresolved doclinks!

Reply:
See https://github.com/firezone/firezone/pull/1605.